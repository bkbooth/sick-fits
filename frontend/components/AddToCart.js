import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

export const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART($itemId: ID!) {
    addToCart(itemId: $itemId) {
      id
      quantity
    }
  }
`;

const AddToCart = ({ itemId }) => (
  <Mutation mutation={ADD_TO_CART_MUTATION} variables={{ itemId }}>
    {(addToCart, { loading }) => (
      <button onClick={addToCart} disabled={loading}>
        ðŸ›’ Add{loading ? 'ing' : ''} to cart
      </button>
    )}
  </Mutation>
);

AddToCart.propTypes = {
  itemId: PropTypes.string.isRequired,
};

export default AddToCart;
