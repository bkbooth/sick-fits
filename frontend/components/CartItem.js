import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import RemoveFromCart from './RemoveFromCart'
import formatMoney from '../lib/formatMoney'

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0;
  }
`

const CartItem = ({ cartItem }) => {
  if (!cartItem.item) return (
    <CartItemStyles>
      <p>This item has been removed.</p>
      <RemoveFromCart id={cartItem.id} />
    </CartItemStyles>
  )
  return <CartItemStyles>
    <img src={cartItem.item.image} width="100" alt={cartItem.item.title} />
    <div className="cart-item-details">
      <h3>{cartItem.item.title}</h3>
      <p>
        {formatMoney(cartItem.item.price * cartItem.quantity)}
        {' - '}
        <em>{cartItem.quantity} Ã— {formatMoney(cartItem.item.price)} each</em>
      </p>
    </div>
    <RemoveFromCart id={cartItem.id} />
  </CartItemStyles>
}

CartItem.propTypes = {
  cartItem: PropTypes.shape({
    id: PropTypes.string,
    quantity: PropTypes.number,
    item: PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      description: PropTypes.string,
      image: PropTypes.string,
      price: PropTypes.number,
    })
  }).isRequired,
}

export default CartItem
