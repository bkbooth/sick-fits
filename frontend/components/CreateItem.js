import React, { Component } from 'react'
import Router from 'next/router'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import Form from './styles/Form'
import formatMoney from '../lib/formatMoney'
import Error from './ErrorMessage'

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
  }

  handleChange = e => {
    const { name, type, value } = e.target
    const val = type === 'number' ? parseFloat(value) : value
    this.setState({ [name]: val })
  }

  uploadFile = async e => {
    const files = e.target.files
    const data = new FormData()
    data.append('file', files[0])
    data.append('upload_preset', 'sick-fits')

    const res = await fetch('https://api.cloudinary.com/v1_1/bkbooth/image/upload', {
      method: 'POST',
      body: data,
    })
    const file = await res.json()

    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    })
  }

  render() {
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
              onSubmit={async e => {
              // Stop the form from submitting
              e.preventDefault()
              // Call the mutation
              const res = await createItem()
              // Redirect user to the single item page
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              })
            }}
            data-test="form"
          >
            <Error error={error} />

            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  required
                  onChange={this.uploadFile}
                />
                {this.state.image && <img src={this.state.image} width="200" alt="Upload Preview" />}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={this.state.title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  required
                  value={this.state.price}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Enter a description"
                  required
                  value={this.state.description}
                  onChange={this.handleChange}
                />
              </label>

              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}

export default CreateItem
export { CREATE_ITEM_MUTATION }
