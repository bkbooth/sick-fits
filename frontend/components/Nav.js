import Link from 'next/link'

const Nav = () => (
  <div>
    <Link href="/">
      <a>Home!</a>
    </Link>
    <Link href="/sell">
      <a>Sell!</a>
    </Link>
  </div>
)

export default Nav
