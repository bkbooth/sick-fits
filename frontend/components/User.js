import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER {
    me {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          title
          price
          image
        }
      }
    }
  }
`;

const User = ({ children }) => {
  const payload = useQuery(CURRENT_USER_QUERY);
  return children(payload);
};

User.propTypes = {
  children: PropTypes.func.isRequired,
};

export default User;
