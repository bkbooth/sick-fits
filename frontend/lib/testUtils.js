import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

export const fakeUser = () => ({
  __typename: 'User',
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
});

export const fakeItem = () => ({
  __typename: 'Item',
  id: casual.uuid,
  title: casual.title,
  description: casual.description,
  price: casual.integer(1000, 30000),
  image: `${casual.word}.jpg`,
  largeImage: `${casual.word}-large.jpg`,
  user: null,
});

export const fakeCartItem = overrides => ({
  __typename: 'CartItem',
  id: casual.uuid,
  quantity: casual.integer(1, 5),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

export function typeInto(wrapper, { inputType = 'input', name, type = 'text', value }) {
  wrapper
    .find(`${inputType}[name="${name}"]`)
    .simulate('change', { target: { name, type, value } });
}
