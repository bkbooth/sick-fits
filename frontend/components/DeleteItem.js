import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import { ALL_ITEMS_QUERY } from 'components/Items';

export const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM($itemId: ID!) {
    deleteItem(id: $itemId) {
      id
    }
  }
`;

const DeleteItem = ({ itemId }) => {
  function createDeleteHandler(mutation) {
    return () =>
      confirm('Are you sure you want to delete this item?') &&
      mutation().catch(error => alert(error.message));
  }

  function cacheUpdate(cache, payload) {
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
  }

  return (
    <Mutation mutation={DELETE_ITEM_MUTATION} variables={{ itemId }} update={cacheUpdate}>
      {(deleteItem, { loading }) => (
        <button onClick={createDeleteHandler(deleteItem)} disabled={loading}>
          🗑️ Delet{loading ? 'ing' : 'e'} item
        </button>
      )}
    </Mutation>
  );
};

DeleteItem.propTypes = {
  itemId: PropTypes.string.isRequired,
};

export default DeleteItem;
