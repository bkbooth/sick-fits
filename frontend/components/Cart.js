import React from 'react'
import { Query, Mutation } from 'react-apollo'
import { adopt } from 'react-adopt'
import gql from 'graphql-tag'
import CartStyles from './styles/CartStyles'
import Supreme from './styles/Supreme'
import CloseButton from './styles/CloseButton'
import SickButton from './styles/SickButton'
import User from './User'
import CartItem from './CartItem'
import TakeMyMoney from './TakeMyMoney'
import calcTotalPrice from '../lib/calcTotalPrice'
import formatMoney from '../lib/formatMoney'
import apostrophiseName from '../lib/apostrophiseName'

const LOCAL_STATE_QUERY = gql`
  query LOCAL_STATE_QUERY {
    cartOpen @client
  }
`

const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`

const Composed = adopt({
  user: ({ render }) => <User>{render}</User>,
  toggleCart: ({ render }) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
})

const Cart = () => (
  <Composed>
    {({ user, toggleCart, localState }) => {
      const me = user.data.me
      if (!me) return null
      return (
        <CartStyles open={localState.data.cartOpen}>
          <header>
            <CloseButton
              title="close"
              onClick={toggleCart}
            >Ã—</CloseButton>
            <Supreme>{apostrophiseName(me.name)} Cart</Supreme>
            <p>You have {me.cart.length} item{me.cart.length === 1 ? '': 's'} in your cart.</p>
          </header>

          <ul>
            {me.cart.map(cartItem => <CartItem cartItem={cartItem} key={cartItem.id} />)}
          </ul>

          <footer>
            <p>{formatMoney(calcTotalPrice(me.cart))}</p>
            <TakeMyMoney>
              <SickButton disabled={!me.cart.length}>Checkout</SickButton>
            </TakeMyMoney>
          </footer>
        </CartStyles>
      )
    }}
  </Composed>
)

export default Cart
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION }
