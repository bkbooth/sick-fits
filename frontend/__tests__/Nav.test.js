import { MockedProvider, wait } from '@apollo/react-testing';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { fakeCartItem, fakeUser } from 'lib/testUtils';
import Nav from 'components/Nav';
import { CURRENT_USER_QUERY } from 'components/User';

const me = fakeUser();

async function render(mocks) {
  const wrapper = mount(
    <MockedProvider mocks={mocks}>
      <Nav />
    </MockedProvider>
  );
  await act(() => wait());
  wrapper.update();
  return wrapper;
}

describe('<Nav />', () => {
  it('renders minimal nav when signed out', async () => {
    const notSignedInMocks = [
      {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: null } },
      },
    ];
    const wrapper = await render(notSignedInMocks);
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(nav).toMatchSnapshot();
    expect(nav.children()).toHaveLength(2);
    expect(nav.find('Link[href="/items"]').exists()).toBe(true);
    expect(nav.find('Link[href="/signup"]').exists()).toBe(true);
  });

  it('renders full nav when signed in', async () => {
    const signedInMocks = [
      {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me } },
      },
    ];
    const wrapper = await render(signedInMocks);
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(nav.children()).toHaveLength(6);
    expect(nav.find('Link[href="/items"]').exists()).toBe(true);
    expect(nav.find('Link[href="/sell"]').exists()).toBe(true);
    expect(nav.find('Link[href="/orders"]').exists()).toBe(true);
    expect(nav.find('Link[href="/me"]').exists()).toBe(true);
    expect(nav.find('[data-test="signout-button"]').exists()).toBe(true);
    expect(nav.find('[data-test="cart-button"]').exists()).toBe(true);
  });

  it('renders the amount of items in the cart', async () => {
    const cart = [fakeCartItem(), fakeCartItem()];
    const signedInWithCartItemsMocks = [
      {
        request: { query: CURRENT_USER_QUERY },
        result: {
          data: { me: { ...me, cart } },
        },
      },
    ];
    const wrapper = await render(signedInWithCartItemsMocks);
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(nav.find('div.count')).toMatchSnapshot();
  });
});
