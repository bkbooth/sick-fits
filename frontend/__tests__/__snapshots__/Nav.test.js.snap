// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders minimal nav when signed out 1`] = `
<ul
  className="NavStyles-sc-11c0d2g-0 fJlgNF"
  data-test="nav"
>
  <Link
    href="/items"
    prefetch={true}
  >
    <a
      href="/items"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/signup"
    prefetch={true}
  >
    <a
      href="/signup"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Signup
    </a>
  </Link>
</ul>
`;

exports[`<Nav /> renders the amount of items in the cart 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-0 jzkIQE"
>
  9
</div>
`;
